### PR TITLE
feat: add rendezvous hashing server selector

### PR DESCRIPTION
**Pull request by cbosss**
Prior to this PR, this library used a modulus based algorithm to determine which server to target for a given key. While modulus is simple and performant, it breaks down when you add/remove servers. In these instances, all keys might need to be migrated to a new servers. This is inefficient and potentially dangerous. 
 
This PR introduces a new selector which uses rendezvous hashing. While this algorithm is slightly more expensive than a simple modulus, it minimizes migrated items when adding/removing servers. This improves the operational characteristics of the library. 

https://en.wikipedia.org/wiki/Rendezvous_hashing

### CODE REVIEW

File: memcache/memcache_test.go [hunk 1]
```diff
@@ -127,7 +127,7 @@ func testWithClient(t *testing.T, c *Client) {
 	if err != ErrMalformedKey {
 		t.Errorf("set(foo bar) should return ErrMalformedKey instead of %v", err)
 	}
-	malFormed = &Item{Key: "foo" + string(0x7f), Value: []byte("foobarval")}
+	malFormed = &Item{Key: "foo" + string(rune(0x7f)), Value: []byte("foobarval")}
```
Comment (cbosss) [inline memcache/memcache_test.go:130]: This is a new warning that was introduced in go 1.13 I believe. Had to fix this up for my local tests to run.

**Comment by cbosss** [comment]
This algorithm is O(n), where n is the number of servers. I think for our usecase that's okay because n is small. However there is a slowdown compared to the old approach. 

```
goos: darwin
goarch: amd64
pkg: github.com/netlify/gomemcache/memcache
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkPickServer
BenchmarkPickServer-8   	32819228	        35.61 ns/op	       0 B/op	       0 allocs/op
PASS

Process finished with the exit code 0
```

```
goos: darwin
goarch: amd64
pkg: github.com/netlify/gomemcache/memcache
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkRendezvousSelector_PickServer
BenchmarkRendezvousSelector_PickServer-8   	 2860376	       424.7 ns/op	       0 B/op	       0 allocs/op
PASS

Process finished with the exit code 0
```

But when compared to making a network call to memcache, i'm not sure nanoseconds matter.

File: memcache/memcache.go [hunk 2]
```diff
@@ -134,18 +134,25 @@ func NewDiscoveryClient(discoveryAddress string, pollingDuration time.Duration)
...
 	}
-	return newDiscoveryClient(discoveryAddress, pollingDuration)
+	return newDiscoveryClient(discoveryAddress, new(ServerList), pollingDuration)
+}
+
+func NewDynamicRendezvousClient(discoveryAddress string, pollingDuration time.Duration) (*Client, error) {
+	// Validate pollingDuration
+	if pollingDuration.Seconds() < 1.0 {
+		return nil, ErrInvalidPollingDuration
+	}
```
Comment (mraerino) [inline memcache/memcache.go]: you don't want to move this check to `newDiscoveryClient`?

File: memcache/selector.go [hunk 3]
```diff
@@ -42,8 +42,16 @@ type ServerSelector interface {
 	// This can be used to get information which is server agnostic.
 	PickAnyServer() (net.Addr, error)
 	Each(func(net.Addr) error) error
+
+	// SetServers updates the list of servers available for selection.
+	SetServers(servers ...string) error
```
Comment (mraerino) [inline memcache/selector.go:47]: does this need to be thread-safe?

File: memcache/selector.go [hunk 4]
```diff
@@ -152,3 +160,49 @@ func (ss *ServerList) PickServer(key string) (net.Addr, error) {
...
+	*ServerList
+}
+
+// PickServer uses teh rendezvous hashing algorith.
+// It loops over each server, hashes key+server together, and selects the server with the largest hash value.
+// Key -> server mappings should be deterministic and uniformly distributed.
+// This method is threadsafe.
+func (rs RendezvousSelector) PickServer(key string) (net.Addr, error) {
+	buf := keyBufPool.Get().(*[]byte)
+	defer keyBufPool.Put(buf)
```
Comment (mraerino) [inline memcache/selector.go:180]: are you sure you need a buffer pool for this? feels like a premature optimization unless you have benchmarked the difference and its big?

File: memcache/memcache.go (hunk 2 above)
Comment (cbosss) [reply memcache/memcache.go]: yeah good call!

File: memcache/selector.go (hunk 3 above)
Comment (cbosss) [reply memcache/selector.go:47]: Yup it does. The `ServerSelector` docs (above) state this:

> All ServerSelector implementations must be safe for concurrent use by multiple goroutines.

File: memcache/selector.go (hunk 4 above)
Comment (cbosss) [reply memcache/selector.go:180]: This just avoids (or I guess amortizes) allocations. The naive checksum+mod implementation was already using this buffer pool, so I thought it was easy enough to add. 

```
BenchmarkRendezvousSelector_PickServer-POOLED		2803006	   433.2 ns/op	   0 B/op	   0 allocs/op

BenchmarkRendezvousSelector_PickServer	               3908797	   308.0 ns/op	   16 B/op	   1 allocs/op
```

I don't have strong opinions!

Comment (mraerino) [reply memcache/selector.go:180]: ah i see. it's fine

File: memcache/selector.go (hunk 3 above)
Comment (mraerino) [reply memcache/selector.go:47]: ah, you're not implementing this yourself. otherwise i would have been confused by the absence of a mutex

Comment (cbosss) [reply memcache/selector.go:47]: Yeah I am just wrapping an existing implementation. The modularity in this lib is nice.

**Review by mraerino — APPROVED** [review]

**Pull request merged**